rfd/0xxx-tls-routing-behind-lb.md

```diff
@@ -0,0 +1,319 @@
+modification does NOT apply to Teleport ALPN protocols that already do mTLS with other certs.
+
+### When to make a connection upgrade
+
+A new **optional** config parameter (e.g. `tsh_routing_upgrade_mode`) will be added in `proxy_service` section in the
+config yaml. This new config has no effect on how Proxy server serve incoming traffic but it will provide a hint to all
+the clients through the webapi ping:
+
+```
+{
+  "auth": {
+  ...
+  },
+  "proxy": {
+  ...
+    "tls_routing_enabled": true,
+    "tls_routing_upgrade_mode": "auto"
+  },
+  ...
+}
+```
+
+`auto` is the default mode where the client performances a test to decide if a connection upgrade is necessary. The test
+is done by sending a TLS handshake with a Teleport custom ALPN to the Proxy server:
+```
+                              ┌───────┐
+                         "on" │upgrade│ "off"
+ ┌────────────────────────────┤ mode  ├────────────────────────────────────┐
+ │                            └──┬────┘                                    │
+ │                               │ "auto" (default)                        │
+ │                               │                                         │
+ │                    ┌──────────▼────────────────┐                        │
+ │                    │     TLS Handshake         │                        │
+ │                    │ALPN teleport-reversetunnel│                        │
+ │                    └────┬─────────────────┬────┘                        │
+ │                         │                 │                             │
+ │                     ┌───▼─────┐        ┌──▼──────┐                      │
+ │                     │Handshake│        │Handshake├───────────────┐      │
+ │ ┌───────────────────┤ Success │        │ Failed  │               │      │
+ │ │                   └───┬─────┘        └┬────────┘               │      │
+ │ │                       │               │                        │      │
+ │ │Negotiated             │No ALPN        │remote error: tls:      │other │
+ │ │teleport-reversetunnel │Negotiated     │no application protocol │errors│
+ │ │                       │               │                        │      │
+ │ │                   ┌───▼────┐          │                  ┌─────▼──┐   │
+ │ │                   │Upgrade ◄──────────┘                  │  Not   │   │
+ └─┼───────────────────►Required│                             │Required◄───┘
+   │                   └────────┘                             └────▲───┘
+   │                                                               │
+   └───────────────────────────────────────────────────────────────┘
+```
+Connection upgrade for TLS Routing should be required when the client and the server fail to neogiate the ALPN protocol,
+hinting there is a load balancer in the middle that terminates TLS.
+
+However, this test may not be bullet proof. In particular, the test explicitly looks for `tls: no application protocol`
```

Comment:
Is there a way to utilize TLS alert codes somehow to determine this? Looks like "no application protocol" alert has code 120, no idea if we can get this info on the client side though:

https://github.com/golang/go/blob/master/src/crypto/tls/alert.go#L50

Reply:
seems `alertNoApplicationProtocol` is only used internally on the TLS server side.